app/controllers/host_factories_controller.rb

```diff
@@ -1,34 +1,45 @@
 # frozen_string_literal: true
 
+# This controller is responsible for creating host records using
+# host factory tokens for authorization.
 class HostFactoriesController < ApplicationController
-  before_filter :verify_token,  only: :create_host
-
   # Ask the host factory to create a host.
-  # This requires the host factory's token in the Authorization header. 
+  # This requires the host factory's token in the Authorization header.
   def create_host
-    id = params.delete(:id)
-    raise ArgumentError, "id" if id.blank?
-
-    @host_builder = HostBuilder.new(@host_factory.account, id, @host_factory.role, @host_factory.role.layers, params)
-    host, api_key = @host_builder.create_host
+    host, api_key = do_create_host
     response = host.as_json
     response['api_key'] = api_key
-    
+
     render json: response, status: :created
   end
-  
+
   protected
-  
-  def verify_token
-    token = request.headers['Authorization'] or raise Unauthorized
-    if token.to_s[/^Token token="(.*)"/]
-      token = $1
-    else
-      raise Unauthorized
+
+  def do_create_host
+    host_factory = hf_token.host_factory
+    role = host_factory.role
+    HostBuilder.new(
+      host_factory.account,
+      params.require(:id),
+      role,
+      role.layers,
+      params.except(:id)
+    ).create_host
+  end
+
+  def hf_token
+    @hf_token ||= HostFactoryToken.from_token(auth_token).tap do |token|
+      raise Unauthorized unless token && [
+        token.valid?,
+        token.valid_origin?(request.ip)
+      ].all?
+    end
+  end
+
+  def auth_token
+    @auth_token ||= begin
```

Comment:
Doesn't look like this needs to be memoized. `def auth_token` is only called once so I don't think you need to set `@auth_token`.

Reply:
You're right, just a habit I guess :) I'll remove it.